node/pkg/p2p/watermark_test.go

```diff
@@ -0,0 +1,192 @@
+		obsvReqSendC:           make(chan *gossipv1.ObservationRequest, cs),
+		sendC:                  make(chan []byte, cs),
+		signedInC:              make(chan *gossipv1.SignedVAAWithQuorum, cs),
+		priv:                   p2ppriv,
+		gk:                     guardianpriv,
+		gst:                    node_common.NewGuardianSetState(nil),
+		nodeName:               nodeName,
+		disableHeartbeatVerify: false,
+		rootCtxCancel:          nil,
+		gov:                    nil,
+		signedGovCfg:           make(chan *gossipv1.SignedChainGovernorConfig, cs),
+		signedGovSt:            make(chan *gossipv1.SignedChainGovernorStatus, cs),
+		components:             DefaultComponents(),
+	}
+
+	// Consume all output channels
+	go func() {
+		name := g.nodeName
+		fmt.Printf("[%s] consuming\n", name)
+		select {
+		case <-g.obsvC:
+			fmt.Printf("[%s] g.obsvC\n", name)
+		case <-g.obsvReqC:
+			fmt.Printf("[%s] g.obsvReqC\n", name)
+		case <-g.signedInC:
+			fmt.Printf("[%s] g.signedInC\n", name)
+		case <-g.signedGovCfg:
+			fmt.Printf("[%s] g.signedGovCfg\n", name)
+		case <-g.signedGovSt:
+			fmt.Printf("[%s] g.signedGovSt\n", name)
+		case <-g.sendC:
+			fmt.Printf("[%s] g.sendC\n", name)
+		}
+	}()
+
+	return g
+}
+
+// TestWatermark runs 4 different guardians one of which does not send its P2PID in the signed part of the heartbeat.
+// The expectation is that hosts that send this information will become "protected" by the Connection Manager.
+func TestWatermark(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create 4 nodes
+	var guardianset = &node_common.GuardianSet{}
+	var gs [4]*G
+	for i, _ := range gs {
+		gs[i] = NewG(fmt.Sprintf("n%d", i))
+		gs[i].components.Port = uint(11000 + i)
+		gs[i].networkID = "/wormhole/localdev"
+
+		guardianset.Keys = append(guardianset.Keys, crypto.PubkeyToAddress(gs[i].gk.PublicKey))
+
+		id, err := p2ppeer.IDFromPublicKey(gs[0].priv.GetPublic())
+		if err != nil {
+			panic(err)
+		}
+
+		gs[i].bootstrapPeers = fmt.Sprintf("/ip4/127.0.0.1/udp/11000/quic/p2p/%s", id.String())
+		gs[i].gst.Set(guardianset)
+
+		gs[i].components.ConnMgr, _ = connmgr.NewConnManager(2, 3, connmgr.WithGracePeriod(2*time.Second))
+	}
+
+	// The 4th guardian does not put its libp2p key in the heartbeat
+	gs[3].components.P2PIDInHeartbeat = false
+
+	// Start the nodes
+	for _, g := range gs {
+		startGuardian(ctx, g)
+	}
+
+	// Wait ~10s to let the nodes gossip.
+	time.Sleep(20 * time.Second)
+
+	// It's expected to have the 3 first nodes protected on every node
+	for gi, g := range gs {
+
+		// expectedProtectedPeers is expected to be 2 for all nodes except the last one where 3 is expected
+		expectedProtectedPeers := 2
+		if gi == 3 {
+			expectedProtectedPeers = 3
+		}
+
+		if len(g.components.ProtectedHostByGuardianKey) != expectedProtectedPeers {
+			t.Errorf("expected to have %d protected peers, got %d", expectedProtectedPeers, len(g.components.ProtectedHostByGuardianKey))
+		}
```

Comment:
standard golang tests are hideous please use testify. Example:

```suggestion
		assert.Equal(len(g.components.ProtectedHostByGuardianKey), expectedProtectedPeers)
```

Or if you want the exact output (you don't need to have it), you can do:

```
assert.Equalf(len(g.components.ProtectedHostByGuardianKey), expectedProtectedPeers, "expected to have %d protected peers, got %d", expectedProtectedPeers, len(g.components.ProtectedHostByGuardianKey)
```